clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,223 @@
+  constexpr const char *l2_output = __builtin_unique_stable_name(l2); // #USN_L2
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_L2{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel2>(l2);
+
+  // kernel3 - expect error
+  // expected-note@#USN_l7{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel3func([]() {}); // #kernel3func_call
+
+  // kernel4 - expect error
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel4func([]() {}); // #kernel4func_call
+
+  // kernel7 - expect error
+  // Same as kernel6, except make l6 part of naming kernel
+  auto l7 = []() { return 1; };
+  auto l8 = [](decltype(l7) *derp = nullptr) { return 2; };
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_l7{{'__builtin_unique_stable_name' evaluated here}}
+  constexpr const char *l7_output = __builtin_unique_stable_name(l7); // #USN_l7
+
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_L2{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel7>(l8);
+
+  // kernel8 and kernel9 - expect no error
+  // Make two lambdas
+  // Call the builtin on the second of the two lambdas (in a constexpr context)
+  // Within a constexpr if, pass the first lambda to a kernel in the true branch,
+  // pass the second lambda to a kernel in the false branch
+  auto l9 = []() { return 1; };
+  auto l10 = []() { return 2; };
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_l10{{'__builtin_unique_stable_name' evaluated here}}
+  constexpr const char *l10_output = __builtin_unique_stable_name(l10); // #USN_l10
+
+  if constexpr (1) {
+    // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+    kernel_single_task<class kernel8>(l9);
+  } else {
+    // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+    // expected-note@#USN_l11{{'__builtin_unique_stable_name' evaluated here}}
+    // expected-note@#USN_l10{{'__builtin_unique_stable_name' evaluated here}}
+    // expected-note@+1{{in instantiation of function template specialization}}
+    kernel_single_task<class kernel9>(l10);
+  }
+
+  // kernel13 and kernel14 - expect no error
+  // pass the same lambda to two kernels
```

Comment:
We do expect errors, at least, we get them in the test case. Also, the passing of the same lambda to two kernels happens within the function call, so it'd be good to clarify that when you update this comment.

Reply:
Restructuring the test to put the kernel uses next to the test case template functions cleared up a lot of mess and the test output makes more sense now!